### PR TITLE
[oneDPL][tests] + Remove support of C++11 and C++14 from oneDPL

### PR DESCRIPTION
**Pull request by SergeyKopienko**
Removed support of C++11 and C++14 from oneDPL library.

### CODE REVIEW

**Comment by SergeyKopienko** [comment]
@MikeDvorskiy could you please take a look to this PR ?

File: include/oneapi/dpl/pstl/onedpl_config.h [hunk 1]
```diff
@@ -168,15 +147,8 @@
...
-#define _ONEDPL_CPP14_2RANGE_MISMATCH_EQUAL_PRESENT                                                                    \
-    (_MSC_VER >= 1900 || __cplusplus >= 201300L || __cpp_lib_robust_nonmodifying_seq_ops == 201304)
-#define _ONEDPL_CPP14_MAKE_REVERSE_ITERATOR_PRESENT                                                                    \
-    (_MSC_VER >= 1900 || __cplusplus >= 201402L || __cpp_lib_make_reverse_iterator == 201402)
-#define _ONEDPL_CPP14_INTEGER_SEQUENCE_PRESENT (_MSC_VER >= 1900 || __cplusplus >= 201402L)
-#define _ONEDPL_CPP14_VARIABLE_TEMPLATES_PRESENT                                                                       \
-    (!__INTEL_COMPILER || __INTEL_COMPILER >= 1700) && (_MSC_FULL_VER >= 190023918 || __cplusplus >= 201402L)
+    (_MSC_VER >= 1912 && _MSVC_LANG >= 201703L) || (_GLIBCXX_RELEASE >= 9 && __GLIBCXX__ >= 20190503)
+#define _ONEDPL_CPP14_VARIABLE_TEMPLATES_PRESENT (!__INTEL_COMPILER || __INTEL_COMPILER >= 1700)
 #define _ONEDPL_CPP11_IS_TRIVIALLY_COPY_ASSIGNABLE_PRESENT                                                             \
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/onedpl_config.h:152]: Do we really need to keep it? Looks that we should it remove to because it is for  "// Required to support GNU libstdc++ below 5.x", which is very old version and I pretty doubt that it works with C++17 compiler,

File: include/oneapi/dpl/pstl/pstl_config.h [hunk 2]
```diff
@@ -99,16 +99,9 @@
...
-#define _PSTL_CPP14_2RANGE_MISMATCH_EQUAL_PRESENT                                                                      \
-    (_MSC_VER >= 1900 || __cplusplus >= 201300L || __cpp_lib_robust_nonmodifying_seq_ops == 201304)
-#define _PSTL_CPP14_MAKE_REVERSE_ITERATOR_PRESENT                                                                      \
-    (_MSC_VER >= 1900 || __cplusplus >= 201402L || __cpp_lib_make_reverse_iterator == 201402)
-#define _PSTL_CPP14_INTEGER_SEQUENCE_PRESENT (_MSC_VER >= 1900 || __cplusplus >= 201402L)
-#define _PSTL_CPP14_VARIABLE_TEMPLATES_PRESENT                                                                         \
-    (!__INTEL_COMPILER || __INTEL_COMPILER >= 1700) && (_MSC_FULL_VER >= 190023918 || __cplusplus >= 201402L)
+    (_MSC_VER >= 1912 && _MSVC_LANG >= 201703L) || (_GLIBCXX_RELEASE >= 9 && __GLIBCXX__ >= 20190503)
+
+#define _PSTL_CPP14_VARIABLE_TEMPLATES_PRESENT (!__INTEL_COMPILER || __INTEL_COMPILER >= 1700)
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/pstl_config.h:104]: Probably we dont need such define as well.

File: include/oneapi/dpl/pstl/onedpl_config.h (hunk 1 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/onedpl_config.h]: Fixed: define removed.

File: include/oneapi/dpl/pstl/pstl_config.h (hunk 2 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/pstl_config.h]: Fixed: define removed.

**Comment by SergeyKopienko** [comment]
I think that we can apply (merge) this PR and discuss about execution policies checks in C++17
in the separate PR  https://github.com/oneapi-src/oneDPL/pull/645

**Comment by MikeDvorskiy** [comment]
> Only C++11 CI tasks are failed now - it's ok because we don't support C++11 and C++14 anymore in oneDPL :
> 
> * https://github.com/oneapi-src/oneDPL/actions/runs/3143407787/jobs/5108115427
> * https://github.com/oneapi-src/oneDPL/actions/runs/3143407787/jobs/5108115824

In that case I believe we should to remove such configurations from CI.

**Comment by SergeyKopienko** [comment]
> > Only C++11 CI tasks are failed now - it's ok because we don't support C++11 and C++14 anymore in oneDPL :
> > 
> > * https://github.com/oneapi-src/oneDPL/actions/runs/3143407787/jobs/5108115427
> > * https://github.com/oneapi-src/oneDPL/actions/runs/3143407787/jobs/5108115824
> 
> In that case I believe we should to remove such configurations from CI.

Corresponding task already created.

File: include/oneapi/dpl/internal/serial_algorithm_impl.h [hunk 3]
```diff
@@ -38,10 +37,9 @@ for_each_n(_InputIterator __first, _Size __n, _Function __f)
     return __first;
 }
 
-#    else
+#else
 using ::std::for_each_n;
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/internal/serial_algorithm_impl.h:41]: Is there any C++ library with doesn't have `::std::for_each_n` and that library is compiled with "std=C++17"?
If there is not, that I think that we should remove 26-42 lines as well.

File: include/oneapi/dpl/internal/serial_numeric_impl.h [hunk 4]
```diff
@@ -17,15 +17,14 @@
...
-#    include <numeric>
+#include <functional>
+#include <iterator>
+#include <numeric>
 namespace oneapi
 {
 namespace dpl
 {
-#    if _ONEDPL_HAS_NUMERIC_SERIAL_IMPL
+#if _ONEDPL_HAS_NUMERIC_SERIAL_IMPL
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/internal/serial_numeric_impl.h:27]: Is there any C++ library with doesn't have `::std::reduce` and that library is compiled with "std=C++17"?
If there is not, that I think that we should remove 27-52 lines as well.

File: include/oneapi/dpl/pstl/onedpl_config.h [hunk 5]
```diff
@@ -27,6 +27,10 @@
 #define ONEDPL_VERSION_MINOR 7
 #define ONEDPL_VERSION_PATCH 1
 
+#if __cplusplus < 201703L
+#    error The oneDPL library requires C++17 or greater
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/onedpl_config.h:31]: "The oneDPL library requires a C++17 compiler."

File: include/oneapi/dpl/execution [hunk 6]
```diff
@@ -18,11 +18,9 @@
 
 #include "oneapi/dpl/internal/common_config.h"
 // ATTENTION!!! Include the header from the C++ standard library before the oneDPL config header to see whether the _PSTL_VERSION macro is defined
-#if __cplusplus >= 201703L
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/execution:21]: Fixed

File: include/oneapi/dpl/pstl/experimental/algorithm [hunk 7]
```diff
@@ -19,16 +19,14 @@
...
-#    if __has_include(<tbb/version.h>)
-#        ifndef PSTL_USE_PARALLEL_POLICIES
-#            define PSTL_USE_PARALLEL_POLICIES (_GLIBCXX_RELEASE != 9)
-#        endif
-#        ifndef _GLIBCXX_USE_TBB_PAR_BACKEND
-#            define _GLIBCXX_USE_TBB_PAR_BACKEND (_GLIBCXX_RELEASE > 10)
-#        endif
-#    endif // __has_include(<tbb/version.h>)
-#endif     // __cplusplus >= 201703L
+#if __has_include(<tbb/version.h>)
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/experimental/algorithm]: Fixed.

File: include/oneapi/dpl/pstl/utils.h [hunk 8]
```diff
@@ -496,46 +493,11 @@ __pstl_left_bound(_Buffer& __a, _Index __first, _Index __last, const _Value& __v
     return __pstl_upper_bound(__a, __beg, __end, __val, __reorder_pred<_Compare>{__comp});
 }
 
-#if (_PSTL_CPP14_INTEGER_SEQUENCE_PRESENT || _ONEDPL_CPP14_INTEGER_SEQUENCE_PRESENT)
-
 template <::std::size_t... _Sp>
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/utils.h]: Fixed.

File: include/oneapi/dpl/internal/gcd_impl.h [hunk 9]
```diff
@@ -54,7 +54,6 @@ __get_abs(_Source __t)
 
 } // namespace internal
 
-#if __cplusplus > 201103L
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/gcd_impl.h:57]: My proposal to discuss about this in separate PR https://github.com/oneapi-src/oneDPL/pull/650

File: include/oneapi/dpl/functional [hunk 10]
```diff
@@ -47,20 +47,13 @@ using ::std::not_equal_to;
...
-#endif // __cplusplus < 201703L
 #if __cplusplus < 202002L
-using ::std::binary_negate;
-using ::std::not1;
-using ::std::not2;
-using ::std::unary_negate;
-#endif // __cplusplus < 202002L
-
-#if __cplusplus > 201103L
+using ::std::binary_negate; // Deprecated in C++17, removed in C++20
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/functional:51]: @ValentinaKats please document next functions as deprecated in our documentation too:
include\oneapi\dpl\functional
 - using ::std::binary_negate; // Deprecated in C++17, removed in C++20
 - using ::std::not1;          // Deprecated in C++17, removed in C++20
 - using ::std::not2;          // Deprecated in C++17, removed in C++20
 - using ::std::unary_negate;  // Deprecated in C++17, removed in C++20

include\oneapi\dpl\type_traits
- using ::std::is_literal_type; // Deprecated in C++17, removed in C++20
- using ::std::result_of;       // Deprecated in C++17, removed in C++20
- using ::std::result_of_t;     // Deprecated in C++17, removed in C++20
- using ::std::is_literal_type_v; // Deprecated in C++17, removed in C++20

File: include/oneapi/dpl/pstl/onedpl_config.h [hunk 11]
```diff
@@ -50,19 +54,8 @@
...
-#    define _DPSTD_DEPRECATED __declspec(deprecated)
-#    define _DPSTD_DEPRECATED_MSG(msg) __declspec(deprecated(msg))
-#elif (__GNUC__ || __clang__)
-#    define _DPSTD_DEPRECATED __attribute__((deprecated))
-#    define _DPSTD_DEPRECATED_MSG(msg) __attribute__((deprecated(msg)))
-#else
-#    define _DPSTD_DEPRECATED
-#    define _DPSTD_DEPRECATED_MSG(msg)
-#endif
+#define _DPSTD_DEPRECATED [[deprecated]]
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/onedpl_config.h:57]: @akukanov what do you think about this? Should we delete this defines too or no?

File: include/oneapi/dpl/pstl/onedpl_config.h (hunk 5 above)
Comment (akukanov) [reply include/oneapi/dpl/pstl/onedpl_config.h:31]: onedpl_config is not a public header to the best of my knowledge, therefore including it and keeping the message consistently shown is under our control I think.

The condition should also include Microsoft's special macro for the language version; as far as I remember, they might not always set `__cplusplus` properly.

As for the wording: first, "the oneDPL library" means "the oneAPI Data Parallel C++ Library library" :) so I would avoid it in our documentation and messages. In this case, I would say "C++ language versions older than C++17 are no more supported by oneDPL" or "oneDPL requires the C++ language version not less than C++17" or "The earliest version of the C++ language that oneDPL supports is C++17", etc.

File: include/oneapi/dpl/functional (hunk 10 above)
[the remark this answers is not in the document]
Comment (akukanov) [reply include/oneapi/dpl/functional:51]: As we discussed offline, since the corresponding aliases already exist in the oneDPL namespace, keeping these intact and declaring deprecated seems more appropriate than removing these. Essentially, we just follow the language: with C++17 it's available but deprecated, in newer versions it's unavailable. Hopefully our documentation, together with the compiler diagnostics about deprecation, will indicate clearly enough that these functions should not be used, neither in kernels nor really anywhere.

File: include/oneapi/dpl/pstl/pstl_config.h [hunk 12]
```diff
@@ -102,14 +102,6 @@
     (_MSC_VER >= 1912 && _MSVC_LANG >= 201703L) ||                                                                     \
         (_GLIBCXX_RELEASE >= 9 && __GLIBCXX__ >= 20190503 && __cplusplus >= 201703L)
 
-#define _PSTL_CPP14_2RANGE_MISMATCH_EQUAL_PRESENT                                                                      \
```
Comment (akukanov) [inline include/oneapi/dpl/pstl/pstl_config.h:105]: The pstl_config file is here only to simplify comparison and sync-up with the upstream configuration (because of different file names); IIRC we keep it as-is with no changes introduced. If we want to clean up this file, we should really do it in the PSTL upstream, not here.

File: include/oneapi/dpl/type_traits [hunk 13]
```diff
@@ -218,10 +215,9 @@ using ::std::is_volatile_v;
 using ::std::negation;
 using ::std::rank_v;
 using ::std::void_t;
-#    if __cplusplus == 201703L
-using ::std::is_literal_type_v;
-#    endif
-#endif // __cplusplus > 201402L
+#if __cplusplus < 202002L
+using ::std::is_literal_type_v; // Deprecated in C++17, removed in C++20
+#endif                          // __cplusplus < 202002L
```
Comment (akukanov) [inline include/oneapi/dpl/type_traits:220]: [[unimportant/ignorable]] I would put all 4 deprecated names next to each other, instead of spreading over the list.

File: include/oneapi/dpl/pstl/onedpl_config.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/onedpl_config.h]: Fixed (removed)

File: include/oneapi/dpl/pstl/onedpl_config.h (hunk 5 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/onedpl_config.h:31]: I would like to use variant "oneDPL requires the C++ language version not less than C++17"
@akukanov, @MikeDvorskiy, @rarutyun are you agree?

File: include/oneapi/dpl/type_traits (hunk 13 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/type_traits]: Fixed.

File: include/oneapi/dpl/pstl/pstl_config.h (hunk 12 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/pstl_config.h:105]: Will investigate in separated task.

File: include/oneapi/dpl/pstl/onedpl_config.h (hunk 5 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/onedpl_config.h]: Also in https://github.com/oneapi-src/oneDPL/pull/642/commits/def98c66d61228b731af1691ccc9272096ed70e7 I add include of onedpl_config.h in each public header file (in some public files it was already included, in other files - no)

Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/onedpl_config.h]: > I would like to use variant "oneDPL requires the C++ language version not less than C++17" @akukanov, @MikeDvorskiy, @rarutyun are you agree?

I dont mind

File: include/oneapi/dpl/cmath [hunk 14]
```diff
@@ -17,7 +17,9 @@
 #define __ONEDPL_cmath
 
 #include "oneapi/dpl/internal/common_config.h"
+// ATTENTION!!! Include the header from the C++ standard library before the oneDPL config header to see whether the _PSTL_VERSION macro is defined
 #include <cmath>
+#include "oneapi/dpl/pstl/onedpl_config.h"
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/cmath:22]: Briefly discussed with Ruslan. It looks like that we have "oneapi/dpl/internal/common_config.h" in each public header..
If so then it make sense to add the diagnostic about "C++17" directly in the "common_config.h", and dont include "onedpl_config.h" in everywhere for that.

File: include/oneapi/dpl/pstl/onedpl_config.h [hunk 15]
```diff
@@ -27,6 +27,10 @@
 #define ONEDPL_VERSION_MINOR 7
 #define ONEDPL_VERSION_PATCH 1
 
+#if __cplusplus < 201703L
```
Comment (rarutyun) [inline include/oneapi/dpl/pstl/onedpl_config.h:30]: seems like everybody agrees that this message is useful. To maximize the changes that user will see this error first I would propose to move it to `common_config` since the latter is available in every public header and goes first.

File: include/oneapi/dpl/optional [hunk 16]
```diff
@@ -17,18 +17,18 @@
 #define __ONEDPL_optional
 
 #include "oneapi/dpl/internal/common_config.h"
+// ATTENTION!!! Include the header from the C++ standard library before the oneDPL config header to see whether the _PSTL_VERSION macro is defined
 #include <optional>
+#include "oneapi/dpl/pstl/onedpl_config.h"
```
Comment (rarutyun) [inline include/oneapi/dpl/optional]: Potentially you discovered an interesting problem. On the other hand, it seems orthogonal to this review. Let's fix it separately.

File: include/oneapi/dpl/cmath (hunk 14 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/cmath]: Fixed.

File: include/oneapi/dpl/pstl/onedpl_config.h (hunk 15 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/onedpl_config.h]: Fixed.

File: include/oneapi/dpl/optional (hunk 16 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/optional]: Fixed: this change was rolled back.

File: include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h [hunk 17]
```diff
@@ -124,30 +124,8 @@ inline device_policy<> dpcpp_default{};
 inline fpga_policy<> dpcpp_fpga{};
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h:121]: Fixed

**Review by rarutyun — APPROVED** [review]
Keeping in mind the CI has passed that patch can be merged

**Comment by SergeyKopienko** [comment]
@MikeDvorskiy could you please take a look at this PR and approve it?

**Review by MikeDvorskiy — APPROVED** [review]
LGTM
